include/nbl/builtin/hlsl/ieee754.hlsl

```diff
@@ -0,0 +1,137 @@
+
+// Copyright (C) 2018-2022 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_HLSL_IEE754_H_INCLUDED_
+#define _NBL_BUILTIN_HLSL_IEE754_H_INCLUDED_
+
+#include <nbl/builtin/hlsl/common.hlsl>
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace numeric_limits
+{
+
+float float_epsilon(float n);
+float float_epsilon(int n);
+float float_epsilon();
+
+}
+
+
+namespace ieee754
+{
+
+uint exponent_bias(in uint exponentBits)
+{
+	return (0x1u << (exponentBits - 1)) - 1;
+}
+uint extract_biased_exponent(float x)
+{
+	return bitfieldExtract(asuint(x), 23, 8);
+}
+int extract_exponent(float x)
+{
+	return int(extract_biased_exponent(x) - exponent_bias(8));
+}
+uint compute_exponent_mask(in uint exponentBits, in uint mantissaBits)
+{
+	return ((1u << exponentBits) - 1) << mantissaBits;
+}
+float replace_biased_exponent(float x, uint exp_plus_bias)
+{
+	return asfloat(bitfieldInsert(asuint(x), exp_plus_bias, 23, 8));
+}
+// performs no overflow tests, returns x*exp2(n)
+float fast_mul_exp2(float x, int n)
+{
+	return replace_biased_exponent(x, extract_biased_exponent(x) + uint(n));
+}
+uint compute_mantissa_mask(in uint mantissaBits)
+{
+	return (0x1u << mantissaBits) - 1;
+}
+uint extract_mantissa(in float x)
+{
+	return (asuint(x) & 0x7fffffu);
+}
+float true_min(in uint exponentBits, in uint mantissaBits)
+{
+	return exp2(1 - int(exponent_bias(exponentBits)) - mantissaBits);
+}
+float min(in uint exponentBits, in uint mantissaBits)
+{
+	const float e = exp2(1 - int(exponent_bias(exponentBits)));
```

Comment:
this stuff should be templated and made conformant with C++ `std::` equivalents and HLSL proposal https://github.com/microsoft/hlsl-specs/blob/main/proposals/0003-numeric-constants.md